p2p/transport/webrtc/datachannel.go

```diff
@@ -0,0 +1,398 @@
+package libp2pwebrtc
+
+import (
+	"context"
+	"io"
+	"net"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-msgio/protoio"
+	"github.com/pion/datachannel"
+	"github.com/pion/webrtc/v3"
+
+	pb "github.com/libp2p/go-libp2p/p2p/transport/webrtc/pb"
+)
+
+var _ network.MuxedStream = &dataChannel{}
+
+const (
+	// maxMessageSize is limited to 16384 bytes in the SDP.
+	maxMessageSize uint64 = 16384
+	// Max message size limit in the SDP is limited to 16384 bytes.
+	// We keep a maximum of 2 messages in the buffer
+	maxBufferedAmount uint64 = 3 * maxMessageSize
+	// bufferedAmountLowThreshold and maxBufferedAmount are bound
+	// to a stream but congestion control is done on the whole
+	// SCTP association. This means that a single stream can monopolize
+	// the complete congestion control window (cwnd) if it does not
+	// read stream data and it's remote continues to send. We can
+	// add messages to the send buffer once there is space for 1 full
+	// sized message.
+	bufferedAmountLowThreshold uint64 = 16384
+
+	protoOverhead  int = 5
+	varintOverhead int = 2
+)
+
+// Package pion detached data channel into a net.Conn
+// and then a network.MuxedStream
+type dataChannel struct {
+	channel       *webrtc.DataChannel
+	rwc           datachannel.ReadWriteCloser
+	laddr         net.Addr
+	raddr         net.Addr
+	readDeadline  time.Time
+	writeDeadline time.Time
+
+	closeWriteOnce sync.Once
+	closeReadOnce  sync.Once
+	resetOnce      sync.Once
+
+	state channelState
+
+	ctx            context.Context
+	cancel         context.CancelFunc
+	m              sync.Mutex
+	readBuf        []byte
+	writeAvailable chan struct{}
+	reader         protoio.Reader
+	writer         protoio.Writer
+
+	requestRead     chan struct{}
+	receivedMessage chan struct{}
+	deadlineUpdated chan struct{}
+
+	wg sync.WaitGroup
+}
+
+func newDataChannel(
+	channel *webrtc.DataChannel,
+	rwc datachannel.ReadWriteCloser,
+	pc *webrtc.PeerConnection,
+	laddr, raddr net.Addr) *dataChannel {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	result := &dataChannel{
+		channel:         channel,
+		rwc:             rwc,
+		laddr:           laddr,
+		raddr:           raddr,
+		readDeadline:    time.Time{},
+		writeDeadline:   time.Time{},
+		ctx:             ctx,
+		cancel:          cancel,
+		writeAvailable:  make(chan struct{}),
+		reader:          protoio.NewDelimitedReader(rwc, 16384),
+		writer:          protoio.NewDelimitedWriter(rwc),
+		readBuf:         []byte{},
+		requestRead:     make(chan struct{}, 5),
```

Comment:
5 is most likely wrong. You need to actually deal with the case where this channel fills up.

Reply:
Actually, this channel will only fill up on concurrent calls to `Read`. I am working under the assumption that `Read` is not safe to call from multiple goroutines. 5 is most likely unnecessary, as the only place where this channel is written to is `Read` until the read side is open on a stream, and `Write`, when the read side is closed. In the latter case, we push to this channel in a non-blocking way which is okay since we only want to receive the muxing protocol flags.